src/infra/core/src/compact_service_impl.rs

```diff
@@ -0,0 +1,289 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::fs;
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+
+use chrono::Utc;
+use datafusion::prelude::*;
+use dill::{component, interface};
+use domain::compact_service::{
+    CompactError,
+    CompactionListener,
+    CompactionMultiListener,
+    CompactionPhase,
+    InvalidDatasetKindError,
+    NullCompactionListener,
+};
+use futures::stream::TryStreamExt;
+use kamu_core::compact_service::CompactService;
+use kamu_core::*;
+use opendatafabric::{
+    DatasetHandle,
+    DatasetKind,
+    DatasetName,
+    MetadataEvent,
+    Multihash,
+    OffsetInterval,
+};
+
+use crate::*;
+
+pub struct CompactServiceImpl {
+    dataset_repo: Arc<dyn DatasetRepository>,
+    dataset_authorizer: Arc<dyn domain::auth::DatasetActionAuthorizer>,
+}
+
+struct LastDatasliceInfo {
+    prev_offset: Option<u64>,
+    new_offset_interval: Option<OffsetInterval>,
+}
+
+struct ChainFilesInfo {
+    data_slice_file_urls: Vec<String>,
+    block_file_urls: Vec<String>,
+    block_hash_to_reset: Multihash,
+    last_dataslice_info: LastDatasliceInfo,
+}
+
+#[component(pub)]
+#[interface(dyn CompactService)]
+impl CompactServiceImpl {
+    pub fn new(
+        dataset_authorizer: Arc<dyn domain::auth::DatasetActionAuthorizer>,
+        dataset_repo: Arc<dyn DatasetRepository>,
+    ) -> Self {
+        Self {
+            dataset_repo,
+            dataset_authorizer,
+        }
+    }
+
+    async fn get_chain_files_info(
+        &self,
+        dataset: Arc<dyn Dataset>,
+    ) -> Result<ChainFilesInfo, CompactError> {
+        let chain = dataset.as_metadata_chain();
+        let head = chain.resolve_ref(&BlockRef::Head).await?;
+        let mut block_hash_to_reset: Option<Multihash> = None;
+        let mut data_slice_file_urls: Vec<String> = vec![];
+        let mut block_file_urls: Vec<String> = vec![];
+
+        let mut block_stream = chain.iter_blocks_interval(&head, None, false);
+        let object_data_repo = dataset.as_data_repo();
+        let object_block_repo = chain.as_metadata_block_repository();
+        let mut last_dataslice_info: Option<LastDatasliceInfo> = None;
+
+        while let Some((block_hash, block)) = block_stream.try_next().await? {
+            block_file_urls.push(object_block_repo.get_internal_url(&block_hash).await.into());
+
+            match block.event {
+                MetadataEvent::AddData(add_data_event) => {
```

Comment:
a) See 'filter_data_stream_blocks', however, be careful, as there are major iteration changes brought by DimaP in his "metadata chain visitor" activities.

b) Why I only see `AddData` for root datasets, and not see `ExecuteTransform` for derived datasets?

Reply:
It was my first solution, But such as I have to take all replaced blocks I decided to rewrite this part to iterate through all blocks to avoid additional iteration during cleaning up